### ISSUE TITLE
Chain of arithmetic ops with Constants should fold into BatchNorm

### ISSUE THREAD

**Issue by vuzelac-cadence** [pytorch/glow#4044]
We have a few disjoint optimizations that we would like to consolidate and upstream.
We are looking to fold a chain of arithmetic operations w/ Constant into a BatchNorm (BN), whenever the chain is preceded by Convolution or BN:
1) If Convolution, fold the chain into BN (This BN is folded into Conv in other optimization).
2) Merge with BatchNorm above the chain (could just create a BN like above, and have a separate step to merge consecutive BNs).

The issue I have is about examining the Constants in RHS/LHS of the arithmetic operations; broadcasts/tiles/constants are already folded into a bigger constant into the importer. So, to examine if the constant is a broadcasted one, we would need to compare it's tensors across the channel dimension.
This computation is relatively demanding and all it does is to annulate the constant folding, as all we do is to extract the original constant. We already have this optimization in a few places and now looking to improve it. However, I just don't see the way out of this excessive computation; we can't import many models/operators w/o the constant folding being enabled in the importer. Do you have any other ideas ?

**Comment by tlepley-cadence** [comment]
@vuzelac-cadence we had an internal discussion and I hope it clarified some of your questions. Should we close this issue ?

**Comment by vuzelac-cadence** [comment]
@tlepley-cadence , our folding is a bit outdated and needs to be improved. Before I work on the version that can be upstreamed, I'd like to hear some feedback.

**Comment by jfix71** [comment]
@vuzelac-cadence I wonder if we should consider only attempting constant folding when we hit an issue where we don't find a Constant. Does that make any sense/would it help at all? That way it's essentially just a fallback for ops that aren't specified nicely with static shapes as we need them, instead of by default trying to constant fold everything.

**Comment by vuzelac-cadence** [comment]
@jfix71 I am not sure I get your question. 
If a whole graph (or subgraph) produces a constant I don't see why would we run it during inference and spend cycles on something that is known. 
I didn't get if your alternative approach is adding to the run time during inference.
I understand it does; if there is e.g. a chain of arithmetic ops with consts, that would not be a candidate for this new "don't fold everything" approach, right ?
Something down the line, in the Backend optimizations, in the target compiler etc simply has to optimize out those constant expressions.

**Comment by jfix71** [comment]
@vuzelac-cadence Sorry, I was unclear -- I meant not always attempt constant folding in the proto loaders specifically. Right now we load the op one at a time, and if it can be constant folded we always do it. However the only reason we do it in the loader is to avoid issues with static shapes. I'm suggesting we only do this constant folding in the loader when necessary (i.e. we search for a Constant and it doesn't exist, try constant folding to see if the Constant can be created).

We would still do constant folding later on in the graph optimizer as well -- #4096 would still work as long as it comes in the pipeline before constant folding.

**Comment by vuzelac-cadence** [comment]
Should have know it's not that simple :) Also, I forgot my own question, now two weeks old.

I am not sure how this search for constant would go; so, we want to avoid folding of Tiles that start from a Constant. That looks exactly the same as the Reshape's Shape case - somewhere up there is a Constant that enabled the folding, right ?
I just don't see how to generically differentiate these cases.  The only approach I see is that the Importer could look for the eligible Arithmetic + Tile + Const sequence and then skips the folding. Same sequence is then to be used in #4096. But since #4096 is ready to go, I am really not looking forward to rework it all.
